0000-experiment-dataframe.md

```diff
@@ -0,0 +1,444 @@
+      name="param1",
+      value=0.1,
+      quality="good",
+      unit="Hz",
+    )
+    artifact1 = ArtifactData(
+      data=mpl_figure,
+      name="curve_plot",
+    )
+    artifact2 = ArtifactData(
+      data=curve_data_table,
+      name="curve_data",
+    )
+    return [result1], [artifact1, artifact2]
+
+```
+
+As you can see, some expected data fields are missing in `result1` in this example (such as component and run time). Note that the minimum required field that the author must provide are name and value. The missing data are implicitly filled by the following base class routine. After this canonicalize, `BaseAnalysis.run` calls `add_analysis_results()` and `add_artifacts()`. The formatted data frame is created (updated) in `ExperimentData` by `add_analysis_results()` call.
+
+### B. Adding new analysis results or artifacts.
+
+An end user (experimentalist) can directly call `add_analysis_results()` and `add_artifacts()` to create custom entry. For example,
+
+```python
+exp_data = MyExperiment(...).run().block_for_result()
+
+exp_data.add_analysis_results(
+  name="my_param1",
+  value="0.1",
+)
+exp_data.add_artifacts(
+  name="my_second_plot",
+  data=my_plot2,
+)
+```
+
+In this case, the user can bypass creation of `AnalysisResult` or `ArtifactData` since these are assumed to be a container used by an experiment author to write `_run_analysis`.
+
+### C. Updating existing results or artifacts.
+
+An end user (experimentalist) can modify analysis result or artifact after experiment.
+
+```python
+exp_data = MyExperiment(...).run().block_for_result()
+
+exp_data.update_analysis_results(
+  index="1fe131b2",
+  tags=["my_project1"],
+)
+exp_data.update_artifacts(
+  index="0ef121s1",
+  data = modified_plot,
+)
+```
+
+Because modified entry must be uniquely specified, `ExperimentData` newly provides methods `update_analysis_results()` and `update_artifacts()` that takes entry ID as a require argument.
+
+### D. Saving results.
+
+Since most end users don't need to explicitly use the service, saving is usually done
+in a transparent manner:
+
+```python
+exp_data = MyExperiment(...).run().block_for_result()
+exp_data.save()
+```
+A specific service can be explicitly given to enable flexibility:
+
+```python
+service = IBMExperimentService(...)
+exp_data = MyExperiment(...).run().block_for_result()
+exp_data.save(service=service)
+```
+
+`save` will return a status object which enables to see which of the save tasks
+succeeded and what errors were raised. Also, `save` can be nonblocking, in which case
+the status object can be used to query whether the save ended and to block:
+
+```python
+exp_data = MyExperiment(...).run().block_for_result()
+save_status = exp_data.save(blocking=False)
+if save_status.running():
+    save_status.block()
+for error in save_status.errors():
+    print(error)
+```
+
+Note that the experiment service already supports API to save artifacts. The payload must be JSON format.
```

Comment:
Should this JSON serialization happen in the service layer, so the python artifict object has a python class as its value? If the later who's responsibility is the JSON serializer in this case? qiskit experiments or the service?

Reply:
Good point. I think this is the responsibility of service since user may want to implement custom (local) service that is capable of handling custom data format. Indeed the conventional IBM provider implements [own encoder](https://github.com/Qiskit/qiskit-ibmq-provider/blob/master/qiskit/providers/ibmq/utils/json_encoder.py) rather than in Qiskit terra. I think we should move current experiment encoder to service.

Reply:
I clarified this in [3de30ba](https://github.com/Qiskit/rfcs/pull/28/commits/3de30ba6235fb53d42881029f2b729cc7dc380e4).